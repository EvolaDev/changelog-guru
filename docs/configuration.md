# Configuration

## Levels of changes

| Default     | CLI Override      | API Override      |
| ----------- | ----------------- | ----------------- |
| _see below_ | `--major <items>` | `major: string[]` |
| _see below_ | `--minor <items>` | `minor: string[]` |
| _see below_ | `--patch <items>` | `patch: string[]` |

Default level of changes:

```YAML
changes:
    major:
        - break
    minor:
        - feat
        - improve
    patch:
        - fix
        - chore
        - refactor
        - test
        - docs
        - build
        - types
        - style
        - workflow
        - perf
        - revert
```

For a list of change types by level, see [SemVer](https://semver.org/). The commits with the specified types will be distributed by change levels. For example:

```

// Commit message with MINOR changes
feat(Core): add awesome feature

```

The following types of changes are defined by default:

-   **MAJOR** version:
    -   `break` - breaking changes
-   **MINOR** version:
    -   `feat` - new features
    -   `improve` - features improvements
-   **PATCH** version:
    -   `fix` - some bugs fixing
    -   `chore` - minor changes
    -   `refactor` - code refactoring
    -   `test` - add or change tests
    -   `docs` - documentation changes
    -   `build` - package changes, release
    -   `types` - code typing
    -   `style` - `css`/`scss`/_other_, style sheet change
    -   `workflow` - workflow changes
    -   `perf` - performance improvements
    -   `revert` - reverted changes

## Output options

Parameters of the output file. Specify the path to the file and the excluded entities.

Default output options:

```YAML
output:
    filePath: CHANGELOG.md
    exclude:
        authorLogin: ['dependabot-preview[bot]']
        commitType: ['build']
        commitScope: ['deps', 'deps-dev']
        commitSubject: ['merge']
```

### filePath

File path to write change log to it.

| Default        | CLI Override      | API Override       |
| -------------- | ----------------- | ------------------ |
| `CHANGELOG.md` | `--output <path>` | `filePath: string` |

### exclude

One way to filter output by ignoring commits with a given type, scope, subject, or from certain authors. To find out about other ways to ignore commits, see the section [Plugins](#plugins)

| Default                                 | CLI Override              | API Override              |
| --------------------------------------- | ------------------------- | ------------------------- |
| _[see output example](#output-options)_ | `--excl-authors<items>`   | `authorLogin: string[]`   |
| _[see output example](#output-options)_ | `--excl-types <items>`    | `commitType: string[]`    |
| _[see output example](#output-options)_ | `--excl-scopes <items>`   | `commitScope: string[]`   |
| _[see output example](#output-options)_ | `--excl-subjects <items>` | `commitSubject: string[]` |

-   **authorLogin** - excludes authors with the listed logins from the output file
-   **commitType** - excludes commits with the listed [types](#commit-structure) from the output file
-   **commitScope** - excludes commits with the listed [scopes](#commit-structure) from the output file
-   **commitSubject** - excludes commits with the listed [subjects](#commit-structure) from the output file

## Other CLI options

| Default                           | CLI             | Description                                                                                                                                      |
| --------------------------------- | --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------ |
| `false`                           | `--bump`        | Based on data about changes made in the project, forms the next version number and bumps it in `package.json`, see [SemVer](https://semver.org/) |
| _current branch from .git folder_ | `--branch`      | Set branch by which change log will be generated                                                                                                 |
| -                                 | `-v, --version` | Show `changelog-guru` package version                                                                                                            |
| -                                 | `--help`        | Show `changelog-guru` cli options help                                                                                                           |

## Plugins

All plugin settings are described in the special section `plugins` of the configuration as follows:

```YAML
plugins:
    <plugin name>:
        <plugins option>
        ...
        <plugins option>
```
