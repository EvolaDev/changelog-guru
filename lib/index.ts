import Reader from './src/reader';

const reader = new Reader();

reader.init();
reader.parse().then(() => {
    process.exit(0);
});
