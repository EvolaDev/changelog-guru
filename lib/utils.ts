import * as fs from 'fs';
import chalk from 'chalk';
import findupSync from 'findup-sync';

export enum Errors {
    GIT_HEAD_IS_NOT_EXIST = '.git/HEAD does not exist',
    GIT_HEAD_SHA_NOT_FOUND = '.git/HEAD ref(s) SHA not found',
    PACKAGE_NOT_FOUND = 'package.json not found',
    CONFIG_NOT_FOUND = '<config> options is not an existing filename',
    TOKEN_IS_NOT_PROVIDED = '<token> options must be provided',
    REPOSITORY_IS_NOT_SPECIFIED = 'project repository not specified',
    PACKAGE_VERSION_IS_INVALID = 'package version is invalid (see https://semver.org/)'
};

export default class Utils {
    static getSHA(cwd: string): string {
        const filepath: string = findupSync('.git/HEAD', { cwd: cwd || process.cwd() });
        let sha: string = "";

        if (fs.existsSync(filepath)) {
            const buffer: Buffer = fs.readFileSync(filepath);
            const match: RegExpExecArray | null = /ref: refs\/heads\/([^\n]+)/.exec(buffer.toString());

            if (match) {
                sha = match[1];
            } else {
                Utils.error(Errors.GIT_HEAD_SHA_NOT_FOUND);
            }
        } else {
            Utils.error(Errors.GIT_HEAD_IS_NOT_EXIST);
        }

        return sha;
    }

    static log(label: string, msg: string) {
        const date = () => new Date().toTimeString().replace(/.*(\d{2}:\d{2}:\d{2}).*/, "$1");

        console.log(chalk`{gray [${date()}]}: {bold ${label}} {greenBright ${msg}}`);
    }

    static info(label: string, msg: string) {
        console.log(chalk`{bold ${label}}: {greenBright ${msg}}`);
    }

    static error(error: Errors) {
        console.error(chalk`{red ${error}}`)

        process.exit(1);
    }
}
