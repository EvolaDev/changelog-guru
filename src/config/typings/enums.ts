export enum ServiceProvider {
    GitHub = 'github',
    GitLab = 'gitlab',
}

export enum ChangeLevel {
    Major = 'major',
    Minor = 'minor',
    Patch = 'patch',
}

export enum ExclusionType {
    AuthorLogin = 'authorLogin',
    CommitType = 'commitType',
    CommitScope = 'commitScope',
    CommitSubject = 'commitSubject',
}
