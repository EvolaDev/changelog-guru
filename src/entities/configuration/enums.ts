export enum ServiceProvider {
    GitHub = 'github',
    GitLab = 'gitlab',
}

export enum Change {
    Major = 'major',
    Minor = 'minor',
    Patch = 'patch',
}

export enum Exclusion {
    Type = 'type',
    Scope = 'scope',
    Subject = 'subject',
    Author = 'author',
}
