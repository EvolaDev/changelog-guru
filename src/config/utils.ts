import { PluginOption, PluginOptionValue } from './typings/types';
import { ValueOf } from '../typings/types';

export class ConfigUtils {
    public static fillFromEnum<T>(options: PluginOption, enumeration: T, mapping: Map<ValueOf<T>, string>): void {
        Object.entries(options).forEach(
            ([name, value]: [string, PluginOption | PluginOption[] | PluginOptionValue | undefined]): void => {
                if (typeof value === 'string') {
                    const type = Object.values(enumeration).find((itemName): boolean => itemName === name);

                    if (value && type) mapping.set(type, value);
                }
            }
        );
    }
}
