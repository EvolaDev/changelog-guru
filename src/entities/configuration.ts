import path from 'path';
import cosmiconfig from 'cosmiconfig';
import { Task } from 'tasktree-cli/lib/task';
import { ServiceProvider } from '../providers/provider';
import { Option, OptionValue } from '../utils/types';
import Key from '../utils/key';
import { Level, FilterType } from '../utils/enums';

export interface Levels extends Option {
    major?: string[];
    minor?: string[];
    patch?: string[];
}

export interface Filters extends Option {
    authors?: string[];
    types?: string[];
    scopes?: string[];
    subjects?: string[];
}

export interface ConfigurationOptions extends Option {
    config?: string;
    levels?: Levels;
    plugins?: string[];
    provider?: ServiceProvider;
    ignore?: Filters;
    [key: string]: Option | OptionValue;
}

export class Configuration {
    public static FILE_NAME = '.changelogrc.yaml';

    private provider: ServiceProvider | undefined;
    private plugins: Set<string> = new Set();
    private levels: Map<string, Level> = new Map();
    private filters: Map<FilterType, string[]> = new Map();
    private options: Option = {};

    public getProvider(): ServiceProvider {
        return this.provider || ServiceProvider.GitHub;
    }

    public getPlugins(): string[] {
        return [...this.plugins.values()];
    }

    public getLevels(): Map<string, Level> {
        return this.levels;
    }

    /*
    public getLevel(type: string): Level {
        return Key.inMap(type, this.levels) || Level.Patch;
    }
    */

    public getFilters(type: FilterType): string[] {
        return this.filters.get(type) || [];
    }

    public getOptions(): ConfigurationOptions {
        return this.options;
    }

    public async load(task: Task): Promise<void> {
        const explorer = cosmiconfig('changelog');
        const userConfig = await explorer.search();
        const defaultConfig = await explorer.load(path.join(__dirname, '../../', Configuration.FILE_NAME));
        let options: ConfigurationOptions;

        if (defaultConfig && !defaultConfig.isEmpty) {
            if (userConfig && !userConfig.isEmpty) {
                options = Object.assign({}, defaultConfig.config, userConfig.config);

                task.log(`Use config file: ${path.relative(process.cwd(), userConfig.filepath)}`);
            } else {
                options = defaultConfig.config;

                task.log(`Use default config file: ${path.relative(process.cwd(), defaultConfig.filepath)}`);
            }

            this.initLevels(options.levels);
            this.initFilters(options.ignore);

            this.provider = options.provider;
            this.plugins = new Set(defaultConfig.config.plugins.concat(options.plugins));
            this.options = new Proxy(options as Option, {
                get(target, name, receiver): Option | undefined {
                    return Reflect.get(target, name, receiver);
                },
            });
        } else {
            task.fail('Default configuration file not found');
        }
    }

    private initLevels(levels?: Levels): void {
        this.levels = new Map();

        if (typeof levels === 'object') {
            const append = (types: string[] | undefined, level: Level): void => {
                if (Array.isArray(types)) {
                    let name: string;

                    types.forEach((type): void => {
                        name = Key.unify(type);

                        if (name.length && !this.levels.has(name)) {
                            this.levels.set(name, level);
                        }
                    });
                }
            };

            append(levels.major, Level.Major);
            append(levels.minor, Level.Minor);
            append(levels.patch, Level.Patch);
        }
    }

    private initFilters(rules?: Filters): void {
        this.filters = new Map();

        if (typeof rules === 'object') {
            const append = (list: string[] | undefined, type: FilterType): void => {
                if (Array.isArray(list)) {
                    this.filters.set(type, [...new Set(list)].filter((filter): boolean => !!Key.unify(filter).length));
                }
            };

            append(rules.authors, FilterType.AuthorLogin);
            append(rules.types, FilterType.CommitType);
            append(rules.scopes, FilterType.CommitScope);
            append(rules.subjects, FilterType.CommitSubject);
        }
    }
}
