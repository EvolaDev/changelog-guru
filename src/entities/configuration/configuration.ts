import * as Enums from './enums';
import * as Types from './types';

export class Configuration {
    public readonly filePath: string;
    public readonly provider: Enums.ServiceProvider;

    private types: Map<string, Enums.Change>;
    private plugins: Map<string, Types.PluginOption>;
    private exclusions: Map<Enums.Exclusion, string[]>;

    public constructor(options: Types.ConfigurationOptions) {
        this.provider = options.provider;
        this.filePath = options.filePath;
        this.types = options.types;
        this.plugins = options.plugins;
        this.exclusions = options.exclusions;
    }

    public getPlugins(): [string, Types.PluginOption][] {
        return [...this.plugins.entries()];
    }

    public getTypes(): [string, Enums.Change][] {
        return [...this.types.entries()];
    }

    public getExclusions(type: Enums.Exclusion): string[] {
        return this.exclusions.get(type) || [];
    }

    /*
    public static fillFromEnum<T>(options: Option, enumeration: T, mapping: Map<ValueOf<T>, string>): void {
        Object.entries(options).forEach(([name, value]: [string, Option | OptionValue]): void => {
            if (typeof value === 'string') {
                const type = Object.values(enumeration).find((itemName): boolean => itemName === name);

                if (value && type) mapping.set(type, value);
            }
        });
    }


*/
}
