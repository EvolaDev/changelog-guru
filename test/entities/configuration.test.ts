import { Task } from 'tasktree-cli/lib/task';
import { Configuration } from '../../src/entities/configuration';
import { FilterType, Level } from '../../src/utils/enums';

describe('Config', (): void => {
    it('Default', (done): void => {
        const config = new Configuration();
        const task = new Task('test configuration loading');

        config.load(task).then((): void => {
            expect(config.getOptions()).toMatchSnapshot();

            expect(config.getFilters(FilterType.AuthorLogin)).toStrictEqual(['dependabot-preview[bot]']);
            expect(config.getFilters(FilterType.CommitType)).toStrictEqual(['build']);
            expect(config.getFilters(FilterType.CommitScope)).toStrictEqual(['deps', 'deps-dev']);
            expect(config.getFilters(FilterType.CommitSubject)).toStrictEqual(['merge']);

            const levels = config.getLevels();

            expect(levels.get('break')).toBe(Level.Major);
            expect(levels.get('feat')).toBe(Level.Minor);
            expect(levels.get('fix')).toBe(Level.Patch);
            expect(levels.get('chore')).toBe(Level.Patch);
            expect(levels.get('refactor')).toBe(Level.Patch);
            expect(levels.get('test')).toBe(Level.Patch);
            expect(levels.get('docs')).toBe(Level.Patch);
            expect(levels.get('build')).toBe(Level.Patch);
            expect(levels.get('types')).toBe(Level.Patch);
            expect(levels.get('style')).toBe(Level.Patch);
            expect(levels.get('workflow')).toBe(Level.Patch);
            expect(levels.get('perf')).toBe(Level.Patch);
            expect(levels.get('revert')).toBe(Level.Patch);

            expect(config.getPlugins()).toStrictEqual(['attention', 'marker', 'scope', 'section']);
            expect(config.getProvider()).toBe('github');
            expect(task.isPending()).toBeTruthy();

            done();
        });
    });
});
