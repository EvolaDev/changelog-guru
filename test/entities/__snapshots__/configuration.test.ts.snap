// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Default 1`] = `
Object {
  "attention": Object {
    "sections": Object {
      "dependencies": "Dependencies",
      "devDependencies": "DevDependencies",
      "engine": "Engine",
      "license": "License",
      "optionalDependencies": "OptionalDependencies,",
      "peerDependencies": "PeerDependencies",
    },
    "templates": Object {
      "added": "Added %name% with %val%",
      "bumped": "Bumped %name% from %pver% to %ver%",
      "changed": "Changed %name% from %pval% to %val%",
      "downgraded": "Downgraded %name% from %pver% to %ver%",
      "removed": "Removed %name%, with %pval%",
      "unchanged": "%name% left unchanged, with %val%",
    },
    "title": "Important Changes",
  },
  "ignore": Object {
    "authors": Array [
      "dependabot-preview[bot]",
    ],
    "scopes": Array [
      "deps",
      "deps-dev",
    ],
    "subjects": Array [
      "merge",
    ],
    "types": Array [
      "build",
    ],
  },
  "levels": Object {
    "major": Array [
      "break",
    ],
    "minor": Array [
      "feat",
      "improve",
    ],
    "patch": Array [
      "fix",
      "chore",
      "refactor",
      "test",
      "docs",
      "build",
      "types",
      "style",
      "workflow",
      "perf",
      "revert",
    ],
  },
  "markers": Object {
    "break": "BREAKING CHANGES",
    "deprecated": "DEPRECATIONS",
    "group": "Grouped commit",
    "hide": "Hidden commits",
    "important": "Important Internal Changes",
  },
  "plugins": Array [
    "attention",
    "marker",
    "scope",
    "section",
  ],
  "provider": "github",
  "scopes": Object {
    "list": Object {
      "api": "API",
      "core": "Core",
      "dts": "TypeScript Declaration Improvements",
      "fc": "Functional Components",
      "ssr": "Server Side Rendering",
    },
    "only": false,
  },
  "sections": Array [
    Object {
      "Features": Array [
        "feat",
      ],
    },
    Object {
      "Improvements": Array [
        "improve",
      ],
    },
    Object {
      "Bug Fixes": Array [
        "fix",
      ],
    },
    Object {
      "Internal changes": Array [
        "types",
        "workflow",
        "build",
        "test",
        "chore",
        "docs",
      ],
    },
    Object {
      "Performance Improvements": Array [
        "perf",
      ],
    },
    Object {
      "Code Refactoring": Array [
        "refactor",
      ],
    },
    Object {
      "Reverts": Array [
        "revert",
      ],
    },
  ],
}
`;
